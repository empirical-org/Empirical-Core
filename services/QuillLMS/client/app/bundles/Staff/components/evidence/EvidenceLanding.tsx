import * as React from "react";
import { Redirect, Route, Switch, withRouter } from 'react-router-dom';

import Activities from './activities';
import Activity from './activity';
import UniversalRulesIndex from './universalRules/universalRules';
import UniversalRule from './universalRules/universalRule';

const EvidenceLanding = () => (
  <div className="main-admin-container">
    <Switch>
      <Redirect exact from='/' to='/activities' />
      <Route component={Activity} path='/activities/:activityId' />
      <Route component={UniversalRule} path='/universal-rules/:ruleId' />
      <Route component={UniversalRulesIndex} path='/universal-rules' />
      <Route component={Activities} path='/activities' />
    </Switch>
  </div>
)

export default withRouter(EvidenceLanding);
