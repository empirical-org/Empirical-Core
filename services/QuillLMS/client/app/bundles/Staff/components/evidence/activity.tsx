import * as React from "react";
import { Redirect, Route, RouteComponentProps, Switch, withRouter } from 'react-router-dom';

import Navigation from './navigation'
import ActivitySettingsWrapper from './configureSettings/activitySettingsWrapper';
import RulesIndexRouter from './configureRules/rulesIndexRouter';
import Rule from './configureRules/rule';
import RulesAnalysis from './rulesAnalysis/rulesAnalysis';
import ActivityStats from './activityStats/activityStats';
import RuleAnalysis from './rulesAnalysis/ruleAnalysis';
import TurkSessions from './gatherResponses/turkSessions';
import SessionsIndex from './activitySessions/sessionsIndex';
import SessionView from './activitySessions/sessionView';
import SemanticLabelsIndex from './semanticRules/semanticLabelsIndex';
import RegexRulesRouter from './regexRules/regexRulesRouter';
import PlagiarismRulesRouter from './plagiarismRules/plagiarismRulesRouter';
import LowConfidenceRulesRouter from './lowConfidenceRules/lowConfidenceRulesRouter';
import ChangeLog from './changeLog/changeLog';
import VersionHistory from "./versionHistory/versionHistory";

import { ActivityRouteProps } from '../../interfaces/evidenceInterfaces';

const Activity: React.FC<RouteComponentProps<ActivityRouteProps>> = ({ match, location, }) => {
  return(
    <React.Fragment>
      <Navigation location={location} match={match} />
      <div className="activity-container">
        <Switch>
          <Redirect exact from='/activities/:activityId' to='/activities/:activityId/settings' />
          <Route component={ActivitySettingsWrapper} path='/activities/new/settings' />
          <Route component={ActivitySettingsWrapper} path='/activities/:activityId/settings' />
          <Route component={Rule} path='/activities/:activityId/rules/:ruleId' />
          <Route component={RulesIndexRouter} path='/activities/:activityId/rules-index' />
          <Route component={ActivityStats} path='/activities/:activityId/stats' />
          <Route component={RuleAnalysis} path='/activities/:activityId/rules-analysis/:promptConjunction/rule/:ruleId/prompt/:promptId' />
          <Route component={RulesAnalysis} path='/activities/:activityId/rules-analysis/:promptConjunction' />
          <Route component={RulesAnalysis} path='/activities/:activityId/rules-analysis' />
          <Route component={TurkSessions} path='/activities/:activityId/turk-sessions' />
          <Route component={SessionView} path='/activities/:activityId/activity-sessions/:sessionId' />
          <Route component={SessionsIndex} path='/activities/:activityId/activity-sessions' />
          <Route component={SemanticLabelsIndex} path='/activities/:activityId/semantic-labels' />
          <Route component={RegexRulesRouter} path='/activities/:activityId/regex-rules' />
          <Route component={PlagiarismRulesRouter} path='/activities/:activityId/plagiarism-rules' />
          <Route component={LowConfidenceRulesRouter} path='/activities/:activityId/low-confidence-rules' />
          <Route component={ChangeLog} path='/activities/:activityId/change-log' />
          <Route component={VersionHistory} path='/activities/:activityId/version-history' />
        </Switch>
      </div>
    </React.Fragment>
  );
}

export default withRouter(Activity)
