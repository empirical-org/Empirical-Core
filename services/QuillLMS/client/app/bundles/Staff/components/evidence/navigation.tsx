import * as React from "react";
import { NavLink, } from 'react-router-dom';

import SubmissionModal from './shared/submissionModal';

const Navigation = ({ location, match }) => {
  const { pathname } = location
  const { params, } = match
  const { activityId, } = params
  const [showSubmissionModal, setShowSubmissionModal] = React.useState<boolean>(false);

  function checkOverviewActive() {
    if(!location) return false;
    return pathname === '/activities';
  }
  function toggleSubmissionModal() {
    setShowSubmissionModal(!showSubmissionModal);
  }

  function renderSubmissionModal () {
    return <SubmissionModal close={toggleSubmissionModal} message='Submission successful!' />;
  }

  let rulesAnalysisSubLinks
  let semanticLabelsSubLinks
  let rulesIndexSubLinks

  if (pathname.includes('rules-analysis')) {
    rulesAnalysisSubLinks = (<React.Fragment>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-analysis/because`}>
        Because
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-analysis/but`}>
        But
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-analysis/so`}>
        So
      </NavLink>
    </React.Fragment>)
  }

  if (pathname.includes('semantic-labels')) {
    semanticLabelsSubLinks = (<React.Fragment>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/semantic-labels/because`}>
        Because
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/semantic-labels/but`}>
        But
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/semantic-labels/so`}>
        So
      </NavLink>
    </React.Fragment>)
  }

  if (pathname.includes('rules-index')) {
    rulesIndexSubLinks = (<React.Fragment>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-index/because`}>
        Because
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-index/but`}>
        But
      </NavLink>
      <NavLink activeClassName="is-active" className="sublink" to={`/activities/${activityId}/rules-index/so`}>
        So
      </NavLink>
    </React.Fragment>)
  }

  let activityEditorAndResults

  if (activityId && activityId !== 'new') {
    activityEditorAndResults = (<React.Fragment>
      <p className="menu-label">
        Activity Editor
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/settings`}>
          Activity Settings
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/semantic-labels`}>
          Semantic Labels
        </NavLink>
        {semanticLabelsSubLinks}
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/regex-rules`}>
          RegEx Rules
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/plagiarism-rules`}>
          Plagiarism Rules
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/low-confidence-rules`}>
          Low Confidence Rules
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/rules-index`}>
          View All Rules
        </NavLink>
        {rulesIndexSubLinks}
      </ul>
      <p className="menu-label">
        Data Generation
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/turk-sessions`}>
          Collect Turk Responses
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/synthetic/seed-data`}>
          Create Seed Data
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/synthetic/labeled-data-upload`}>
          Generate Labeled Synthetic Data
        </NavLink>
      </ul>
      <p className="menu-label">
        Activity Results
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/stats`}>
          Activity Stats
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/activity-sessions`}>
          View Sessions
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/rules-analysis`}>
          Rules Analysis
        </NavLink>
        {rulesAnalysisSubLinks}
      </ul>
      <p className="menu-label">
        Change Log
      </p>
      <ul className="menu-list">
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/change-log`}>
          Change Log
        </NavLink>
        <NavLink activeClassName="is-active" to={`/activities/${activityId}/version-history`}>
          Version History
        </NavLink>
      </ul>
    </React.Fragment>)
  }

  return(
    <React.Fragment>
      <section className="left-side-menu">
        <p className="menu-label">
          Tool Views
        </p>
        <ul className="menu-list">
          <NavLink activeClassName='is-active' isActive={checkOverviewActive} to="/activities">
            View Activities
          </NavLink>
          <NavLink activeClassName='is-active' to="/activities/new">
            Create New Activity
          </NavLink>
          <NavLink activeClassName='is-active' to="/universal-rules">
            View Universal Rules
          </NavLink>
          <NavLink activeClassName='is-active' to="/health-dashboard">
            Health Dashboard
          </NavLink>
        </ul>
        {activityEditorAndResults}
      </section>
      {showSubmissionModal && renderSubmissionModal()}
    </React.Fragment>
  )
}

export default Navigation
