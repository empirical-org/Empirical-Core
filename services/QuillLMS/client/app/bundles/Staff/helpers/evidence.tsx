import stripHtml from "string-strip-html";

import {
  MINIMUM_READING_LEVEL,
  MAXIMUM_READING_LEVEL,
  TARGET_READING_LEVEL,
  SCORED_READING_LEVEL,
  IMAGE_LINK,
  IMAGE_ALT_TEXT,
  IMAGE_CAPTION,
  IMAGE_ATTRIBUTION,
  HIGHLIGHT_PROMPT,
  PLAGIARISM,
  FLAG
} from '../../../constants/evidence';
import { DropdownObjectInterface } from '../interfaces/evidenceInterfaces'

export const buildActivity = ({
  activityFlag,
  activityNotes,
  activityTitle,
  activityScoredReadingLevel,
  activityTargetReadingLevel,
  activityParentActivityId,
  activityPassages,
  activityMaxFeedback,
  activityBecausePrompt,
  activityButPrompt,
  activitySoPrompt,
  highlightPrompt,
}) => {
  const prompts = [activityBecausePrompt, activityButPrompt, activitySoPrompt];
  const maxFeedback = activityMaxFeedback || 'Nice effort! You worked hard to make your sentence stronger.';
  prompts.forEach(prompt => prompt.max_attempts_feedback = maxFeedback);
  return {
    activity: {
      notes: activityNotes,
      title: activityTitle,
      parent_activity_id: activityParentActivityId ? parseInt(activityParentActivityId) : null,
      flag: activityFlag,
      scored_level: activityScoredReadingLevel,
      target_level: parseInt(activityTargetReadingLevel),
      highlight_prompt: highlightPrompt,
      passages_attributes: activityPassages,
      prompts_attributes: prompts
    }
  };
}

const targetReadingLevelError = (value: string) => {
  if(!value) {
    return `${TARGET_READING_LEVEL} must be a number between 4 and 12, and cannot be blank.`
  }
  const num = parseInt(value);
  if(num < MINIMUM_READING_LEVEL || num > MAXIMUM_READING_LEVEL) {
    return `${TARGET_READING_LEVEL} must be a number between ${MINIMUM_READING_LEVEL} and ${MAXIMUM_READING_LEVEL}.`
  }
}

const scoredReadingLevelError = (value: string) => {
  if(value === '') {
    return;
  }
  const num = parseInt(value);
  if(isNaN(num) || num < 4 || num > 12) {
    return `${SCORED_READING_LEVEL} must be a number between 4 and 12, or left blank.`;
  }
}

export const handlePageFilterClick = ({
  startDate,
  endDate,
  turkSessionID,
  filterOption,
  setStartDate,
  setEndDate,
  setShowError,
  setPageNumber,
  setTurkSessionIDForQuery,
  setFilterOptionForQuery,
  storageKey }: {
    startDate: Date,
    endDate?: Date,
    filterOption?: DropdownObjectInterface,
    turkSessionID?: string,
    setStartDate: (startDate: string) => void,
    setEndDate: (endDate: string) => void,
    setTurkSessionIDForQuery?: (turkSessionID: string) => void,
    setFilterOptionForQuery?: (filterOption: DropdownObjectInterface) => void,
    setShowError: (showError: boolean) => void,
    setPageNumber: (pageNumber: DropdownObjectInterface) => void,
    storageKey: string,
  }) => {
  if(!startDate) {
    setShowError(true);
    return;
  }
  setShowError(false);
  setPageNumber && setPageNumber({ value: '1', label: "Page 1" })
  const startDateString = startDate.toISOString();
  window.sessionStorage.setItem(`${storageKey}startDate`, startDateString);
  setStartDate(startDateString);
  if(!endDate) {
    // reset to null when user has cleared endDate value
    setEndDate(null);
  } else if(endDate)  {
    const endDateString = endDate.toISOString();
    window.sessionStorage.setItem(`${storageKey}endDate`, endDateString);
    setEndDate(endDateString);
  }
  if(turkSessionID === '') {
    // reset to null so backend doesn't check on empty string
    setTurkSessionIDForQuery(null);
  } else if(turkSessionID) {
    window.sessionStorage.setItem(`${storageKey}turkSessionId`, turkSessionID);
    setTurkSessionIDForQuery(turkSessionID);
  }
  if(filterOption) {
    window.sessionStorage.setItem(`${storageKey}filterOption`, JSON.stringify(filterOption));
    setFilterOptionForQuery(filterOption);
  }
}

export const validateForm = (keys: string[], state: any[], ruleType?: string) => {
  let errors = {};
  state.map((value, i) => {
    switch(keys[i]) {
      case IMAGE_LINK:
      case IMAGE_ALT_TEXT:
      case IMAGE_CAPTION:
      case IMAGE_ATTRIBUTION:
      case HIGHLIGHT_PROMPT:
      case FLAG:
        break;
      case TARGET_READING_LEVEL:
        const targetError = targetReadingLevelError(value);
        if(targetError) {
          errors[TARGET_READING_LEVEL] = targetError;
        }
        break;
      case SCORED_READING_LEVEL:
        const scoredError = scoredReadingLevelError(value);
        if(scoredError) {
          errors[SCORED_READING_LEVEL] = scoredError
        }
        break;
      case "Stem Applied":
        const stemsApplied = Object.keys(value).filter(stem => value[stem].checked);
        if(!stemsApplied.length) {
          errors[keys[i]] = 'You must select at least one stem.';
        }
        if(ruleType && ruleType === PLAGIARISM && stemsApplied.length > 1) {
          errors[keys[i]] = 'You can only select one stem.';
        }
        break;
      case "Concept UID":
        if(!value) {
          errors[keys[i]] = 'Concept UID cannot be blank. Default for plagiarism rules is "Kr8PdUfXnU0L7RrGpY4uqg"'
        }
        break;
      default:
        const strippedValue = value && stripHtml(value);
        if(!strippedValue || strippedValue.length === 0) {
          errors[keys[i]] = `${keys[i]} cannot be blank.`;
        }
     }
  });
  return errors;
}

export function titleCase(string: string){
  return string[0].toUpperCase() + string.slice(1).toLowerCase();
}
