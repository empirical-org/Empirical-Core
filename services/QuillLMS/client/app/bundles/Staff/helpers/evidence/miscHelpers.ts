import stripHtml from "string-strip-html";

import {
  MINIMUM_READING_LEVEL,
  MAXIMUM_READING_LEVEL,
  TARGET_READING_LEVEL,
  SCORED_READING_LEVEL,
  IMAGE_LINK,
  IMAGE_ALT_TEXT,
  IMAGE_CAPTION,
  IMAGE_ATTRIBUTION,
  HIGHLIGHT_PROMPT,
  PLAGIARISM,
  FLAG,
  TEXT,
  BUILDING_ESSENTIAL_KNOWLEDGE,
  MAX_ATTEMPTS_FEEDBACK,
  HIGHLIGHTING_PROMPT,
  IMAGE,
  PROMPTS,
  BREAK_TAG
} from '../../../../constants/evidence';
import { DEFAULT_HIGHLIGHT_PROMPT } from "../../../Shared";
import { DropdownObjectInterface, PromptInterface } from '../../interfaces/evidenceInterfaces'
import { getCheckIcon } from "./renderHelpers";

export const buildActivity = ({
  activityFlag,
  activityNotes,
  activityTitle,
  activityParentActivityId,
  activityPassages,
  activityBecausePrompt,
  activityButPrompt,
  activitySoPrompt,
  highlightPrompt,
}) => {
  const prompts = [activityBecausePrompt, activityButPrompt, activitySoPrompt];
  return {
    activity: {
      notes: activityNotes,
      title: activityTitle,
      parent_activity_id: activityParentActivityId ? parseInt(activityParentActivityId) : null,
      flag: activityFlag,
      highlight_prompt: highlightPrompt,
      passages_attributes: activityPassages,
      prompts_attributes: prompts
    }
  };
}

const targetReadingLevelError = (value: string) => {
  if(!value) {
    return `${TARGET_READING_LEVEL} must be a number between 4 and 12, and cannot be blank.`
  }
  const num = parseInt(value);
  if(num < MINIMUM_READING_LEVEL || num > MAXIMUM_READING_LEVEL) {
    return `${TARGET_READING_LEVEL} must be a number between ${MINIMUM_READING_LEVEL} and ${MAXIMUM_READING_LEVEL}.`
  }
}

const scoredReadingLevelError = (value: string) => {
  if(value === '') {
    return;
  }
  const num = parseInt(value);
  if(isNaN(num) || num < 4 || num > 12) {
    return `${SCORED_READING_LEVEL} must be a number between 4 and 12, or left blank.`;
  }
}

export const handlePageFilterClick = ({
  startDate,
  endDate,
  filterOption,
  setStartDate,
  setEndDate,
  setShowError,
  setPageNumber,
  setFilterOptionForQuery,
  storageKey }: {
    startDate: Date,
    endDate?: Date,
    filterOption?: DropdownObjectInterface,
    setStartDate: (startDate: string) => void,
    setEndDate: (endDate: string) => void,
    setFilterOptionForQuery?: (filterOption: DropdownObjectInterface) => void,
    setShowError: (showError: boolean) => void,
    setPageNumber: (pageNumber: DropdownObjectInterface) => void,
    storageKey: string,
  }) => {
  if(!startDate) {
    setShowError(true);
    return;
  }
  setShowError(false);
  setPageNumber && setPageNumber({ value: '1', label: "Page 1" })
  const startDateString = startDate.toISOString();
  window.sessionStorage.setItem(`${storageKey}startDate`, startDateString);
  setStartDate(startDateString);
  if(!endDate) {
    // reset to null when user has cleared endDate value
    setEndDate(null);
  } else if(endDate)  {
    const endDateString = endDate.toISOString();
    window.sessionStorage.setItem(`${storageKey}endDate`, endDateString);
    setEndDate(endDateString);
  }
  if(filterOption) {
    window.sessionStorage.setItem(`${storageKey}filterOption`, JSON.stringify(filterOption));
    setFilterOptionForQuery(filterOption);
  }
}

export const validateForm = (keys: string[], state: any[], ruleType?: string) => {
  let errors = {};
  state.map((value, i) => {
    switch(keys[i]) {
      case IMAGE_LINK:
      case IMAGE_ALT_TEXT:
      case IMAGE_CAPTION:
      case IMAGE_ATTRIBUTION:
      case HIGHLIGHT_PROMPT:
      case FLAG:
        break;
      case TARGET_READING_LEVEL:
        const targetError = targetReadingLevelError(value);
        if(targetError) {
          errors[TARGET_READING_LEVEL] = targetError;
        }
        break;
      case SCORED_READING_LEVEL:
        const scoredError = scoredReadingLevelError(value);
        if(scoredError) {
          errors[SCORED_READING_LEVEL] = scoredError
        }
        break;
      case "Stem Applied":
        const stemsApplied = Object.keys(value).filter(stem => value[stem].checked);
        if(!stemsApplied.length) {
          errors[keys[i]] = 'You must select at least one stem.';
        }
        if(ruleType && ruleType === PLAGIARISM && stemsApplied.length > 1) {
          errors[keys[i]] = 'You can only select one stem.';
        }
        break;
      case "Concept UID":
        if(!value) {
          errors[keys[i]] = 'Concept UID cannot be blank. Default for plagiarism rules is "Kr8PdUfXnU0L7RrGpY4uqg"'
        }
        break;
      default:
        const strippedValue = value && stripHtml(value);
        if(!strippedValue || strippedValue.length === 0) {
          errors[keys[i]] = `${keys[i]} cannot be blank.`;
        }
    }
  });
  return errors;
}

export function validateFormSection({
  label,
  activityPassages,
  activityBecausePrompt,
  activityButPrompt,
  activitySoPrompt,
}) {
  switch(label) {
    case titleCase(TEXT):
      const passagePresent = activityPassages && activityPassages[0] && activityPassages[0].text && activityPassages[0].text !== BREAK_TAG;
      return getCheckIcon(passagePresent);
    case BUILDING_ESSENTIAL_KNOWLEDGE:
      const essentialKnowledgePresent = (
        activityPassages && activityPassages[0] &&
        !!activityPassages[0].essential_knowledge_text &&
        activityPassages[0].essential_knowledge_text !== BREAK_TAG
      );
      return getCheckIcon(essentialKnowledgePresent);
    case HIGHLIGHTING_PROMPT:
      const highlightingPresent = (activityPassages && activityPassages[0] && !!activityPassages[0].highlight_prompt && activityPassages[0].highlight_prompt !== DEFAULT_HIGHLIGHT_PROMPT);
      return getCheckIcon(highlightingPresent);
    case IMAGE:
      const imageDetailsPresent = (
        activityPassages &&
        activityPassages[0] &&
        !!activityPassages[0].image_link &&
        !!activityPassages[0].image_alt_text &&
        !!activityPassages[0].image_caption &&
        !!activityPassages[0].image_attribution
      );
      return getCheckIcon(imageDetailsPresent);
    case MAX_ATTEMPTS_FEEDBACK:
      const maxAttemptsFeedbackPresent = (
        activityBecausePrompt &&
        activityButPrompt &&
        activitySoPrompt &&
        !!activityBecausePrompt.max_attempts_feedback &&
        !!activityButPrompt.max_attempts_feedback &&
        !!activitySoPrompt.max_attempts_feedback &&
        activityBecausePrompt.max_attempts_feedback !== BREAK_TAG &&
        activityButPrompt.max_attempts_feedback !== BREAK_TAG &&
        activitySoPrompt.max_attempts_feedback !== BREAK_TAG
      );
      return getCheckIcon(maxAttemptsFeedbackPresent);
    case PROMPTS:
      const promptsDetailsPresent = (
        activityBecausePrompt &&
        activityButPrompt &&
        activitySoPrompt &&
        !!activityBecausePrompt.text &&
        !!activityButPrompt.text &&
        !!activitySoPrompt.text &&
        !!activityBecausePrompt.first_strong_example &&
        !!activityButPrompt.first_strong_example &&
        !!activitySoPrompt.first_strong_example &&
        !!activityBecausePrompt.second_strong_example &&
        !!activityButPrompt.second_strong_example &&
        !!activitySoPrompt.second_strong_example
      );
      return getCheckIcon(promptsDetailsPresent);
    default:
      break
  }
}

export function titleCase(string: string){
  return string[0].toUpperCase() + string.slice(1).toLowerCase();
}
