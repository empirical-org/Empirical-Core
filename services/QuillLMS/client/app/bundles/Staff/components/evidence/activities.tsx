import * as React from "react";
import { useQuery } from 'react-query';
import { Link } from 'react-router-dom';

import Navigation from './navigation';

import { DataTable, Error, FlagDropdown, Spinner } from '../../../Shared/index';
import { getCheckIcon, renderErrorsContainer } from "../../helpers/evidence/renderHelpers";
import { ActivityInterface } from '../../interfaces/evidenceInterfaces';
import { fetchActivities } from '../../utils/evidence/activityAPIs';

const Activities = ({ location, match }) => {

  // cache activity data for updates
  const { data: activitiesData } = useQuery("activities", fetchActivities);
  const [errors, setErrors] = React.useState<string[]>([])
  const [flag, setFlag] = React.useState<string>('alpha')

  const filteredActivities = activitiesData && activitiesData.activities && activitiesData.activities.filter(act => flag === 'All Flags' || act.flag === flag) || []

  const formattedRows = filteredActivities.map((activity: ActivityInterface) => {
    const { id, title, invalid_highlights, parent_activity_id, notes } = activity;
    const activityLink = (<Link to={`/activities/${id}`}>{title}</Link>);
    const highlightLabel = (<Link to={`/activities/${id}`}>{getCheckIcon(!(invalid_highlights && invalid_highlights.length))}</Link>);
    return {
      id,
      parent_activity_id,
      title: activityLink,
      notes,
      valid_highlights: highlightLabel
    }
  });

  function onFlagChange(e) { setFlag(e.target.value) }

  if(!activitiesData) {
    return(
      <React.Fragment>
        <Navigation location={location} match={match} />
        <div className="loading-spinner-container">
          <Spinner />
        </div>
      </React.Fragment>
    );
  }

  if(activitiesData.error) {
    return(
      <div className="error-container">
        <Error error={`${activitiesData.error}`} />
      </div>
    );
  }

  const dataTableFields = [
    { name: "Parent Activity ID", attribute:"parent_activity_id", width: "100px" },
    { name: "Internal Name", attribute:"notes", width: "450px" },
    { name: "Name", attribute:"title", width: "350px" },
    { name: "Highlight Validation", attribute:"valid_highlights", width: "100px", noTooltip: true }
  ];

  return(
    <React.Fragment>
      <Navigation location={location} match={match} />
      {errors && renderErrorsContainer(false, errors)}
      <div className="activities-container">
        <FlagDropdown flag={flag} handleFlagChange={onFlagChange} isLessons={true} />
        <DataTable
          className="activities-table"
          defaultSortAttribute="name"
          headers={dataTableFields}
          rows={formattedRows ? formattedRows : []}
        />
      </div>
    </React.Fragment>
  );
}

export default Activities
