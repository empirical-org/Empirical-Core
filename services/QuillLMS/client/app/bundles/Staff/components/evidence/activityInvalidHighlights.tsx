import * as React from "react";
import { useQuery } from 'react-query';
import { Link } from 'react-router-dom';

import { Spinner, } from '../../../Shared/index';
import { getCheckIcon, } from "../../helpers/evidence/renderHelpers";
import { fetchInvalidHighlights, } from '../../utils/evidence/activityAPIs';

const ActivityInvalidHighlights = ({ activityId }) => {

  const { data: invalidHighlights } = useQuery({
    queryKey: ['invalidHighlight', activityId],
    queryFn: fetchInvalidHighlights
  })

  if (!invalidHighlights) {
    return (
      <span></span>
    )
  }

  return (
    <Link to={`/activities/${activityId}`}>
      {getCheckIcon(!(invalidHighlights.invalid_highlights && invalidHighlights.invalid_highlights.length))}
    </Link>
  )
}

export default ActivityInvalidHighlights
