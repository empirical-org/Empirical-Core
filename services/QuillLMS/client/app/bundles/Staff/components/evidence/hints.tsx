import * as React from "react";
import { useQuery, useQueryClient, } from 'react-query';
import { useHistory, } from 'react-router-dom';

import Navigation from './navigation';

import { HintInterface } from '../../interfaces/evidenceInterfaces';
import { createHint, deleteHint, fetchHints, updateHint, } from '../../utils/evidence/hintAPIs';
import { DropdownInput, Error, Spinner, Snackbar, defaultSnackbarTimeout, } from '../../../Shared/index';
import { renderErrorsContainer } from "../../helpers/evidence/renderHelpers";
import RuleHint from "./configureRules/ruleHint";
import RuleHintDropdown from "./configureRules/ruleHintDropdown";

const Hints = ({ location, match }) => {
  const { params } = match;
  const { hintId } = params;

  const history = useHistory();

  const newHint: HintInterface = {
    id: ''
    name: '',
    explanation: '',
    image_link: 'no image',
    image_alt_text: ''
  }

  // cache hint data for updates
  const { data: hintsData } = useQuery("hints", fetchHints);
  const [errors, setErrors] = React.useState<string[]>([])
  const [hint, setHint] = React.useState<HintInterface>(newHint)
  const [snackbarText, setSnackbarText] = React.useState<string[]>('')
  const [snackbarVisible, setSnackbarVisible] = React.useState<Boolean>(false)

  const queryClient = useQueryClient()

  if (hintId && hintsData && hint.id != hintId) {
    const hintFromUrl = hintsData.hints.find((hint) => hint.id == hintId)
    if (hintFromUrl) setHint(hintFromUrl)
  }

  const onHintChange = (selectedHint) => {
    if (!selectedHint) {
      return setHint(newHint)
    }
    setHint(selectedHint)
    history.push({
      pathname: `/hints/${selectedHint.id}`
    })
  }

  const handleSaveHint = () => {
    if (hint.id == '') {
      createHint(hint).then(handleSaveResponse)
    } else {
      updateHint(hint.id, hint).then(handleSaveResponse)
    }
  }

  const handleSaveResponse = (response) => {
    if (response.errors.length > 0) setErrors(result.errors)

    queryClient.refetchQueries("hints")
    if (response.hint) {
      setHint(response.hint)
    }
  }

  const handleDeleteHint = () => {
    if (!window.confirm('Are you sure you want to delete this Hint?  Doing so will affect all Rules that it is attached to.')) return

    flashSnackbar('Hint deleted.')
    deleteHint(hint.id).then(handleSaveResponse)
      .then(() => setHint(newHint))
  }

  const flashSnackbar = (text: string) => {
    setSnackbarText(text)
    setSnackbarVisible(true)
    setTimeout(() => setSnackbarVisible(false), defaultSnackbarTimeout)
  }

  if(!hintsData) {
    return(
      <React.Fragment>
        <Navigation location={location} match={match} />
        <div className="loading-spinner-container">
          <Spinner />
        </div>
      </React.Fragment>
    );
  }

  if(hintsData.error) {
    return(
      <div className="error-container">
        <Error error={`${hintsData.error}`} />
      </div>
    );
  }

  const createNewHintOption = (
    <option value={''}>Create New Hint</option>
  )

  const hintDropdown = () => {
    return (
      <RuleHintDropdown
        emptySelectionText="Create new hint"
        onHintChange={onHintChange}
        selectedHintId={hint.id}
      />
    )
  }

  const hintForm = () => {
    return (
      <RuleHint
        errors={errors}
        hint={hint}
        setHint={setHint}
      />
    )
  }

  const saveButton = () => {
    const buttonText = hint.id == '' ? 'Create Hint' : 'Update Hint'
    return (<button className="quill-button medium primary outlined save-hint" onClick={handleSaveHint} type="button">{buttonText}</button>)
  }

  const deleteButton = () => {
    if (hint.id == '') return;

    return (<button className="quill-button medium primary outlined delete-hint" onClick={handleDeleteHint} type="button">Delete Hint</button>)
  }
    
  return(
    <React.Fragment>
      <Navigation location={location} match={match} />
      <div className="hints-container">
        <h3>Rule Hints</h3>
        {hintDropdown()}
        {errors && renderErrorsContainer(true, errors)}
        <div className="hint-form">
          {hintForm()}
          <div className="button-wrapper">
            {saveButton()}
            {deleteButton()}
          </div>
        </div>
        <Snackbar text={snackbarText} visible={snackbarVisible} />
      </div>
    </React.Fragment>
  );
}

export default Hints
