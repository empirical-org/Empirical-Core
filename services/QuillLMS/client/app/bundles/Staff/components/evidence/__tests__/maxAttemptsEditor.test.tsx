import * as React from 'react';
import { shallow } from 'enzyme';

import MaxAttemptsEditor from '../configureSettings/maxAttemptsEditor';
import { BECAUSE, FEEDBACK } from '../../../../../constants/evidence';

const mockProps = {
  conjunction: '',
  prompt: {
    conjunction: BECAUSE,
    text: '1',
    max_attempts: 5,
    max_attempts_feedback: FEEDBACK,
    first_strong_example: '',
    second_strong_example: ''
  },
  handleSetPrompt: jest.fn()
}

describe('MaxAttemptsEditor component', () => {
  const container = shallow(<MaxAttemptsEditor {...mockProps} />);

  it('should render MaxAttemptsEditor', () => {
    expect(container).toMatchSnapshot();
  });
});
