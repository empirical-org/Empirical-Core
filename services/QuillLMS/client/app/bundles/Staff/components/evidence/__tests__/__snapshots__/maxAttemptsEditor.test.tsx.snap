// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxAttemptsEditor component should render MaxAttemptsEditor 1`] = `
<div>
  <p
    className="text-editor-label has-text"
  >
     - Max Attempts Feedback - Student Did Not Reach Optimal AutoML Label
  </p>
  <TextEditor
    ContentState={[Function]}
    EditorState={[Function]}
    handleTextChange={[Function]}
    key="but-max-attempt-feedback"
    shouldCheckSpelling={true}
    text="feedback"
  />
</div>
`;
